### PR TITLE
WIP mb/image2pulsetrain

### PR DESCRIPTION
**Pull request by mbeyeler**
Work in progress: Add functionality to encode an image as a list of pulse trains.

### CODE REVIEW

File: pulse2percept/electrode2currentmap.py [hunk 1]
```diff
@@ -570,6 +597,121 @@ def get_pulse(pulse_dur, tsample, interphase_dur, pulsetype):
...
+    pulsetype: {'cathodicfirst', 'anodicfirst'}, optional
+
+    """
+
+    # TODO: Default RF size could be half of the electrode gap?
+    def __init__(self, img, implant, coding='amplitude', valrange=[0, 50],
+                 const_amp=20, const_freq=20, rftype='gaussian', rfsize=None,
+                 tsample=0.005 / 1000, dur=0.5, pulsedur=0.5 / 1000.,
+                 interphasedur=0.5 / 1000., pulsetype='cathodicfirst'):
+        try:
```
Comment (arokem) [inline pulse2percept/electrode2currentmap.py]: Maybe only try to import skimage if a string is provided, or if the image is RGB? We could allow a simple gray-scale input, if people want to provide that directly.

File: pulse2percept/electrode2currentmap.py [hunk 2]
```diff
@@ -570,6 +597,121 @@ def get_pulse(pulse_dur, tsample, interphase_dur, pulsetype):
...
+        assert len(valrange) == 2 and valrange[1] > valrange[0]
+
+        if isinstance(img, str):
+            # Load image from filename
+            img_large = imread(img, as_grey=True).astype(np.float32)
+        else:
+            # Convert image to grayscale
+            img_large = rgb2gray(np.array(img)).astype(np.float32)
+
+        # Maximize contrast
```
Comment (arokem) [inline pulse2percept/electrode2currentmap.py]: Wouldn't this lead to confusing results in an experiment in which different contrasts are used?

File: pulse2percept/electrode2currentmap.py [hunk 3]
```diff
@@ -570,6 +597,121 @@ def get_pulse(pulse_dur, tsample, interphase_dur, pulsetype):
...
+
+        # Center the image on the implant
+        xyr = np.array([[e.x_center, e.y_center, e.radius] for e in implant])
+        xlo = np.min(xyr[:, 0] - xyr[:, 2])
+        xhi = np.max(xyr[:, 0] + xyr[:, 2])
+        ylo = np.min(xyr[:, 1] - xyr[:, 2])
+        yhi = np.max(xyr[:, 1] + xyr[:, 2])
+
+        # Resize the image accordingly and flip up down
+        img_resize = resize(img_large, (yhi - ylo, xhi - xlo))
```
Comment (arokem) [inline pulse2percept/electrode2currentmap.py]: I guess we'll always need skimage for this...

File: pulse2percept/electrode2currentmap.py [hunk 4]
```diff
@@ -570,6 +597,121 @@ def get_pulse(pulse_dur, tsample, interphase_dur, pulsetype):
...
+        yg = np.linspace(ylo, yhi, img_resize.shape[0])
+        xg = np.linspace(xlo, xhi, img_resize.shape[1])
+        yg, xg = np.meshgrid(yg, xg)
+
+        # For each electrode, find the stimulation strength (magnitude)
+        magn = []
+        for e in implant:
+            rf = receptive_field(e, xg, yg, rftype, rfsize)
+            magn.append(np.mean(rf * img_resize))
+        magn = np.array(magn)
```
Comment (arokem) [inline pulse2percept/electrode2currentmap.py]: We might be able to preallocate this one.

File: pulse2percept/electrode2currentmap.py [hunk 5]
```diff
@@ -570,6 +597,121 @@ def get_pulse(pulse_dur, tsample, interphase_dur, pulsetype):
...
+            elif coding == 'frequency':
+                # Map magnitude to frequency
+                amp = const_amp
+                freq = m
+            else:
+                e_s = "Acceptable values for `coding` are 'amplitude' or"
+                e_s += "'frequency'."
+                raise ValueError(e_s)
+
+            pt = Psycho2Pulsetrain(tsample, freq=freq, amp=amp, dur=dur,
```
Comment (arokem) [inline pulse2percept/electrode2currentmap.py]: Maybe there's a way to speed up/save memory by doing these all in the same time-series object? OTOH, might not be worth the hassle.

File: pulse2percept/electrode2currentmap.py (hunk 2 above)
Comment (mbeyeler) [inline pulse2percept/electrode2currentmap.py]: Good point. Added an input argument `max_contrast` so users can choose.

File: pulse2percept/electrode2currentmap.py (hunk 3 above)
Comment (mbeyeler) [inline pulse2percept/electrode2currentmap.py]: Yup, unfortunately...

File: pulse2percept/electrode2currentmap.py (hunk 1 above)
Comment (mbeyeler) [inline pulse2percept/electrode2currentmap.py]: Agree, except we always need it for resizing the image accordingly... (see below)

**Pull request closed without merging**